src/main/java/com/github/firmwehr/gentle/firm/optimization/EscapeAnalysisOptimization.java

```diff
@@ -0,0 +1,392 @@
+			return Optional.of(allocationCall);
+		}
+		return Optional.empty();
+	}
+
+	private boolean escapes(Proj callResProj) {
+		Deque<Node> deepOuts = new ArrayDeque<>();
+		deepOuts.add(callResProj);
+		while (!deepOuts.isEmpty()) {
+			Node next = deepOuts.removeLast();
+			for (BackEdges.Edge edge : BackEdges.getOuts(next)) {
+				if (edge.node.getMode().equals(Mode.getM())) {
+					LOGGER.debug("stop at %s (memory)", edge.node);
+					continue;
+				}
+				if (edge.node instanceof Call || edge.node instanceof Return) {
+					LOGGER.debug("%s escapes on %s", callResProj, edge.node);
+					return true;
+				}
+				if (edge.node instanceof Load load) {
+					if (fromPredIfMatches(load.getPtr()).isPresent()) {
+						// Load from this call => ignore (would be replaced)
+						LOGGER.debug("stop at %s", edge.node);
+						continue;
+					}
+					LOGGER.debug("%s escapes on %s (invalid address)", callResProj, edge.node);
+					return true;
+				}
+				if (edge.node instanceof Store store) {
+					if (!callResProjIsReachableFromStoreValue(store.getValue(), callResProj, new HashSet<>())) {
+						if (fromPredIfMatches(store.getPtr()).isPresent()) {
+							// Store *in* allocated object would be replaced
+							LOGGER.debug("stop at %s", edge.node);
+							continue;
+						}
+					}
+					// a) store in other object, mark this as escape. If other object does not escape, this
+					//    can be optimized when running escape analysis again
+					// b) the store address is in an unknown format, we can't rewrite it.
+					//    In that case, we mark it as escaped
+					LOGGER.debug("%s escapes on %s", callResProj, edge.node);
+					return true;
+				}
+				if (edge.node instanceof Cmp) {
+					// we don't have a pointer to compare, but this should be optimized away
+					// in a different step anyway
+					LOGGER.debug("%s is used in %s", next, edge.node);
+					return true;
+				}
+				if (edge.node instanceof Proj proj && proj.getMode().equals(Mode.getM())) {
+					// don't follow memory nodes
+					LOGGER.debug("stop at %s", edge.node);
+					continue;
+				}
+				// a node that can't be ignored
+				deepOuts.add(edge.node);
+			}
+		}
+		LOGGER.info("%s does not escape, can be optimized", callResProj);
+		return false;
+	}
+
+	private boolean callResProjIsReachableFromStoreValue(Node storePred, Proj callResProj, Set<Node> seen) {
+		if (!seen.add(storePred)) {
+			return false;
+		}
+		if (storePred.equals(callResProj)) {
+			return true;
+		}
+		for (Node pred : storePred.getPreds()) {
+			// load does not escape
+			// store would have been checked separately already
+			if (pred instanceof Load || pred instanceof Store) {
+				continue;
+			}
+			// memory nodes don't lead to callResProj (as it is a value)
+			if (!pred.getMode().equals(Mode.getM()) && callResProjIsReachableFromStoreValue(pred, callResProj, seen)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	static class CallRewriter {
+
+		private final Call call;
```

Comment:
Probably a good idea to put some inline comments on these to explain what they are used for